src/compute-client/src/controller.rs

```diff
@@ -1155,11 +1141,16 @@ where
                 .collection_mut(*id)
                 .expect("Reference to absent collection");
 
-            collection.write_frontier.join_assign(new_upper);
+            collection
+                .write_frontier_upper
+                .join_assign(&new_upper.upper);
+            collection
+                .write_frontier_lower
+                .join_assign(&new_upper.lower);
```

Comment:
Here would be a good place to enforce that `write_frontier_lower >= write_frontier_upper + some_maximum_lag`, to ensure that compaction is not held back arbitrarily far. However, the compute controller only deals in logical times, so there is no way to specify `some_maximum_lag` in terms of an actual time duration.